### PR TITLE
client-go/features: warn when ordering initialization issue

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
SetFeatureGates logs a warning when the default env var
implementation has been already used.
Such a situation indicates a potential ordering issue and usually is unwanted.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
requires https://github.com/kubernetes/kubernetes/pull/122555
please review only the second commit

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/features/envvar.go [hunk 1]
```diff
@@ -73,6 +77,10 @@ type envVarFeatureGates struct {
 	// readEnvVarsOnce guards reading environmental variables
 	readEnvVarsOnce sync.Once
 
+	// readEnvVars holds the boolean value which
+	// indicates whether readEnvVarsOnce has been called.
+	readEnvVars *atomic.Value
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/features/envvar.go:82]: atomic.Bool

File: staging/src/k8s.io/client-go/features/features.go [hunk 2]
```diff
@@ -99,8 +101,23 @@ func AddFeaturesToExistingFeatureGates(registry Registry) error {
 //	// then replace client-go's feature gates implementation with your implementation
 //	clientgofeaturegate.SetFeatureGates(utilfeature.DefaultMutableFeatureGate)
 func SetFeatureGates(newFeatureGates Gates) {
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/features/features.go:103]: Stefan requested a rename to `ReplaceFeatureGates` I think.

Comment (p0lyn0mial) [reply staging/src/k8s.io/client-go/features/features.go:103]: thanks, I missed it.

File: staging/src/k8s.io/client-go/features/features.go [hunk 3]
```diff
@@ -99,8 +101,23 @@ func AddFeaturesToExistingFeatureGates(registry Registry) error {
...
+	if setFeatureGatesWithWarningIndicator(newFeatureGates) {
+		klog.Warningf("The default feature gates implementation has already been used and now it's being overwritten. This might lead to unexpected behaviour. Check your initialization order.")
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply staging/src/k8s.io/client-go/features/features.go:105]: isn't it for errors ?

File: staging/src/k8s.io/client-go/features/envvar.go (hunk 1 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/client-go/features/envvar.go:82]: a cargo cult, same applies to `atomic.Value`

**Comment by p0lyn0mial** [comment]
/close

pulled into https://github.com/kubernetes/kubernetes/pull/122555

**Pull request closed without merging**